pkg/network/dhcp/configurator_test.go

```diff
@@ -0,0 +1,174 @@
+		dhcpConfigurationFactory = fake.NewFakeInMemoryNetworkCacheFactory()
+	})
+
+	BeforeEach(func() {
+		// make sure the test can write a file in the whatever dir ensure uses.
+		Expect(os.MkdirAll(filepath.Dir(kubevirtPrivatePath), 0755)).To(Succeed())
+	})
+
+	AfterEach(func() {
+		Expect(os.RemoveAll(kubevirtPrivatePath)).To(Succeed())
+	})
+
+	Context("dhcp configurator with client filtering", func() {
+		BeforeEach(func() {
+			addr, err := netlink.ParseAddr(advertisingCIDR)
+			Expect(err).NotTo(HaveOccurred())
+			advertisingAddr = *addr
+
+			dhcpConfigurator = NewConfiguratorWithClientFilter(dhcpConfigurationFactory, launcherPID, bridgeName, fakeHandler)
+		})
+
+		It("should fail when nothing to load", func() {
+			_, err := dhcpConfigurator.ImportConfiguration(ifaceName)
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("should succeed when cache file present", func() {
+			dhcpConfig := cache.DhcpConfig{
+				Name:         ifaceName,
+				IP:           netlink.Addr{},
+				Mtu:          1400,
+				IPAMDisabled: false,
+			}
+			Expect(dhcpConfigurator.ExportConfiguration(dhcpConfig)).To(Succeed())
+
+			readConfig, err := dhcpConfigurator.ImportConfiguration(ifaceName)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(*readConfig).To(Equal(dhcpConfig))
+		})
+
+		Context("start DHCP function", func() {
+			var dhcpConfig cache.DhcpConfig
+			var dhcpOptions *v1.DHCPOptions
+
+			BeforeEach(func() {
+				dhcpConfig = cache.DhcpConfig{
+					Name:              ifaceName,
+					IP:                netlink.Addr{},
+					AdvertisingIPAddr: advertisingAddr.IP,
+					Mtu:               1400,
+					IPAMDisabled:      false,
+				}
+			})
+
+			It("should succeed when DHCP server started", func() {
+				fakeHandler.EXPECT().StartDHCP(&dhcpConfig, advertisingAddr.IP, bridgeName, nil, true).Return(nil)
+
+				Expect(dhcpConfigurator.EnsureDhcpServerStarted(ifaceName, dhcpConfig, dhcpOptions)).To(Succeed())
+			})
+
+			It("should fail when DHCP server failed", func() {
+				fakeHandler.EXPECT().StartDHCP(&dhcpConfig, advertisingAddr.IP, bridgeName, nil, true).Return(fmt.Errorf("failed to start DHCP server"))
+
+				Expect(dhcpConfigurator.EnsureDhcpServerStarted(ifaceName, dhcpConfig, dhcpOptions)).To(HaveOccurred())
+			})
+
+			When("IPAM is disabled on the DhcpConfig", func() {
+				BeforeEach(func() {
+					dhcpConfig = cache.DhcpConfig{
+						Name:              ifaceName,
+						IP:                netlink.Addr{},
+						AdvertisingIPAddr: advertisingAddr.IP,
+						Mtu:               1400,
+						IPAMDisabled:      true,
+					}
+				})
+
+				It("should never actually try to start the DHCP server", func() {
+					fakeHandler.EXPECT().StartDHCP(&dhcpConfig, advertisingAddr.IP, bridgeName, nil, true).Return(nil).Times(0)
+
+					Expect(dhcpConfigurator.EnsureDhcpServerStarted(ifaceName, dhcpConfig, dhcpOptions)).To(Succeed())
+				})
+			})
+		})
+	})
+
+	Context("dhcp configurator without client filtering", func() {
```

Comment:
The test in this section are very similar to the tests in the previos one. Can `DescribeTable` be used?

Reply:
This is at context level; if I used a table, I'd lose the `BeforeEach` blocks that compose the test setup, which would severely impact the test's readability. I prefer it as is.

Furthermore, the tests are composed of 2 lines (one setting up the expected mock, another issuing the call) I don't see much to be gained from using a table.

Reply:
What is the thumb rule to decide what should be tested in the first section and what in the second?
For example the second section lacks those two tests -
"IPAM is disabled on the DhcpConfig"
"should succeed when DHCP server is started multiple times".

If you don't want to use a `DescribeTable` I would just have one or two tests testing the `filtering` `no filtering`. And the rest using  one of the options (`filtering` or not). I wouldn't duplicate all (most of) the tests.

Reply:
Everything should be unit tested. If using a table is the cost, so be it.